sql/analyzer/apply_foreign_keys.go

```diff
@@ -160,57 +165,249 @@ func applyForeignKeysToNodes(ctx *sql.Context, a *Analyzer, n sql.Node, cache *f
+			newDeclaredFks = append(newDeclaredFks, declaredFks[i+1:]...)
+			declaredFks = newDeclaredFks
+		}
+	}
+	// If there are no foreign keys (or we've ignored them all) then we can skip the rest of this
+	if len(declaredFks) == 0 {
+		return nil, nil
+	}
+
+	fkEditor := &plan.ForeignKeyEditor{
+		Schema:     tbl.Schema(),
+		Editor:     nil,
+		References: make([]*plan.ForeignKeyReferenceHandler, len(declaredFks)),
+		RefActions: nil,
+	}
+	for i, declaredFk := range declaredFks {
+		// Tables do not include their database. As a workaround, we'll use one of foreign keys to tell us the database.
+		if editor == nil {
+			editor, err = cache.Add(ctx, tbl, declaredFk.Database, declaredFk.Table)
+			if err != nil {
+				return nil, err
+			}
+			fkEditor.Editor = editor
+			fkChain = fkChain.AddTable(declaredFk.ReferencedDatabase, declaredFk.ReferencedTable)
+		}
+
+		refTbl, refUpdater, err := cache.Get(ctx, a, declaredFk.ReferencedDatabase, declaredFk.ReferencedTable)
+		if err != nil {
+			return nil, err
+		}
+		refIndex, ok, err := plan.FindIndexWithPrefix(ctx, refTbl, declaredFk.ReferencedColumns)
+		if err != nil {
+			return nil, err
+		}
+		if !ok {
+			// If this error is returned, it is due to an index deletion not properly checking for foreign key usage
+			//TODO: enforce that the last matching index cannot be removed if depended upon by a foreign key
+			return nil, fmt.Errorf("no suitable index found for table `%s` in foreign key `%s` declared on table `%s`",
+				declaredFk.ReferencedTable, declaredFk.Name, declaredFk.Table)
+		}
+		indexPositions, appendTypes, err := plan.FindForeignKeyColMapping(ctx, declaredFk.Name, tbl, declaredFk.Columns,
+			declaredFk.ReferencedColumns, refIndex)
+		if err != nil {
+			return nil, err
+		}
+		var selfCols map[string]int
+		if declaredFk.IsSelfReferential() {
+			selfCols = make(map[string]int)
+			for i, col := range tbl.Schema() {
+				selfCols[strings.ToLower(col.Name)] = i
+			}
+		}
+		fkEditor.References[i] = &plan.ForeignKeyReferenceHandler{
+			ForeignKey: declaredFk,
+			SelfCols:   selfCols,
+			RowMapper: plan.ForeignKeyRowMapper{
+				Index:          refIndex,
+				Updater:        refUpdater,
+				SourceSch:      tbl.Schema(),
+				IndexPositions: indexPositions,
+				AppendTypes:    appendTypes,
+			},
+		}
+	}
+	return fkEditor, nil
+}
+
+// getForeignKeyRefActions returns an editor containing only the referential actions to enforce on the given table.
+func getForeignKeyRefActions(ctx *sql.Context, a *Analyzer, tbl sql.ForeignKeyTable, cache *foreignKeyTableCache, fkChain foreignKeyChain) (*plan.ForeignKeyEditor, error) {
+	var editor sql.ForeignKeyUpdater
+	referencedFks, err := tbl.GetReferencedForeignKeys(ctx)
+	if err != nil {
+		return nil, err
+	}
+	// MySQL has a CASCADE limit of a depth of 15
+	//TODO: figure out if MySQL errors once 15 has been hit, or if it just stops cascading
+	if len(referencedFks) == 0 || fkChain.Count() >= 15 {
+		return nil, nil
```

Comment:
We should error here for now, just to be safe

Reply:
To clarify the `TODO`, it's not an error here during analysis, but rather if a `CASCADE` actually hits a depth of 15 at runtime. So if it does error at 15, then I'd replace the `ForeignKeyEditor` node with a `ForeignKeyDepthError` node that just returns an error.